backend/src/schema/resolvers/reports.spec.js

```diff
@@ -1,521 +1,593 @@
+          beforeEach(async () => {
+            await factory.create('Post', {
+              author: currentUser,
+              id: 'post-to-report-id',
+              title: 'This is a post that is going to be reported',
+              categoryIds,
+            })
           })
-          variables = {
-            ...variables,
-            resourceId: 'p23',
-          }
-        })
 
-        it('returns type "Post"', async () => {
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              type: 'Post',
-            },
+          it('returns type "Post"', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: {
+                  ...variables,
+                  resourceId: 'post-to-report-id',
+                },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  type: 'Post',
+                },
+              },
+              errors: undefined,
+            })
           })
-        })
 
-        it('returns resource in post attribute', async () => {
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              post: {
-                title: 'Matt and Robert having a pair-programming',
+          it('returns resource in post attribute', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: {
+                  ...variables,
+                  resourceId: 'post-to-report-id',
+                },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  post: {
+                    title: 'This is a post that is going to be reported',
+                  },
+                },
               },
-            },
+              errors: undefined,
+            })
           })
-        })
 
-        it('returns null in user attribute', async () => {
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              user: null,
-            },
+          it('returns null in user attribute', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: {
+                  ...variables,
+                  resourceId: 'post-to-report-id',
+                },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  user: null,
+                },
+              },
+              errors: undefined,
+            })
           })
         })
-      })
-
-      /* An der Stelle würde ich den p23 noch mal prüfen, diesmal muss aber eine error meldung kommen.
-           At this point I would check the p23 again, but this time there must be an error message. */
 
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-21T08:22:56Z" title="Monday, October 21st 2019, 10:22:56 am +02:00">Oct 21, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -1,521 +1,599 @@
-import { GraphQLClient } from 'graphql-request'
-import Factory from '../../seed/factories'
-import { host, login, gql } from '../../jest/helpers'
-import { getDriver, neode } from '../../bootstrap/neo4j'
 import { createTestClient } from 'apollo-server-testing'
 import createServer from '../.././server'
+import Factory from '../../seed/factories'
+import { gql } from '../../jest/helpers'
+import { getDriver, neode as getNeode } from '../../bootstrap/neo4j'
 
 const factory = Factory()
-const instance = neode()
+const instance = getNeode()
 const driver = getDriver()
 
-describe('report mutation', () => {
-  let reportMutation
-  let headers
-  let client
-  let variables
-  let createPostVariables
-  let user
+describe('report resources', () => {
+  let authenticatedUser, currentUser, mutate, query, moderator, abusiveUser
   const categoryIds = ['cat9']
-
-  const action = () => {
-    reportMutation = gql`
-      mutation($resourceId: ID!, $reasonCategory: ReasonCategory!, $reasonDescription: String!) {
-        report(
-          resourceId: $resourceId
-          reasonCategory: $reasonCategory
-          reasonDescription: $reasonDescription
-        ) {
-          createdAt
-          reasonCategory
-          reasonDescription
-          type
-          submitter {
-            email
-          }
-          user {
-            name
-          }
-          post {
-            title
-          }
-          comment {
-            content
-          }
+  const reportMutation = gql`
+    mutation($resourceId: ID!, $reasonCategory: ReasonCategory!, $reasonDescription: String!) {
+      report(
+        resourceId: $resourceId
+        reasonCategory: $reasonCategory
+        reasonDescription: $reasonDescription
+      ) {
+        createdAt
+        reasonCategory
+        reasonDescription
+        type
+        submitter {
+          email
+        }
+        user {
+          name
+        }
+        post {
+          title
+        }
+        comment {
+          content
         }
       }
-    `
-    client = new GraphQLClient(host, {
-      headers,
-    })
-    return client.request(reportMutation, variables)
+    }
+  `
+  const variables = {
+    resourceId: 'whatever',
+    reasonCategory: 'other',
+    reasonDescription: 'Violates code of conduct !!!',
   }
 
-  beforeEach(async () => {
-    variables = {
-      resourceId: 'whatever',
-      reasonCategory: 'other',
-      reasonDescription: 'Violates code of conduct !!!',
-    }
-    headers = {}
-    user = await factory.create('User', {
-      id: 'u1',
-      role: 'user',
-      email: 'test@example.org',
-      password: '1234',
-    })
-    await factory.create('User', {
-      id: 'u2',
-      role: 'user',
-      name: 'abusive-user',
-      email: 'abusive-user@example.org',
-    })
-    await instance.create('Category', {
-      id: 'cat9',
-      name: 'Democracy & Politics',
-      icon: 'university',
+  beforeAll(() => {
+    const { server } = createServer({
+      context: () => {
+        return {
+          driver,
+          neode: instance,
+          user: authenticatedUser,
+        }
+      },
     })
+    mutate = createTestClient(server).mutate
+    query = createTestClient(server).query
   })
 
   afterEach(async () => {
     await factory.cleanDatabase()
   })
 
-  describe('unauthenticated', () => {
-    it('throws authorization error', async () => {
-      await expect(action()).rejects.toThrow('Not Authorised')
-    })
-  })
-
-  describe('authenticated', () => {
-    beforeEach(async () => {
-      headers = await login({
-        email: 'test@example.org',
-        password: '1234',
+  describe('report a resource', () => {
+    describe('unauthenticated', () => {
+      it('throws authorization error', async () => {
+        authenticatedUser = null
+        await expect(mutate({ mutation: reportMutation, variables })).resolves.toMatchObject({
+          data: { report: null },
+          errors: [{ message: 'Not Authorised!' }],
+        })
       })
     })
 
-    describe('invalid resource id', () => {
-      it('returns null', async () => {
-        await expect(action()).resolves.toEqual({
-          report: null,
+    describe('authenticated', () => {
+      beforeEach(async () => {
+        currentUser = await factory.create('User', {
+          id: 'current-user-id',
+          role: 'user',
+          email: 'test@example.org',
+          password: '1234',
+        })
+        await factory.create('User', {
+          id: 'abusive-user-id',
+          role: 'user',
+          name: 'abusive-user',
+          email: 'abusive-user@example.org',
         })
+        await instance.create('Category', {
+          id: 'cat9',
+          name: 'Democracy & Politics',
+          icon: 'university',
+        })
+
+        authenticatedUser = await currentUser.toJson()
       })
-    })
 
-    describe('valid resource id', () => {
-      describe('reported resource is a user', () => {
-        beforeEach(async () => {
-          variables = {
-            ...variables,
-            resourceId: 'u2',
-          }
+      describe('invalid resource id', () => {
+        it('returns null', async () => {
+          await expect(mutate({ mutation: reportMutation, variables })).resolves.toMatchObject({
+            data: { report: null },
+            errors: undefined,
+          })
         })
+      })
 
-        it('returns type "User"', async () => {
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              type: 'User',
-            },
+      describe('valid resource', () => {
+        describe('reported resource is a user', () => {
+          it('returns type "User"', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: { ...variables, resourceId: 'abusive-user-id' },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  type: 'User',
+                },
+              },
+              errors: undefined,
+            })
           })
-        })
 
-        it('returns resource in user attribute', async () => {
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              user: {
-                name: 'abusive-user',
+          it('returns resource in user attribute', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: { ...variables, resourceId: 'abusive-user-id' },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  user: {
+                    name: 'abusive-user',
+                  },
+                },
               },
-            },
+              errors: undefined,
+            })
           })
-        })
 
-        it('returns the submitter', async () => {
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              submitter: {
-                email: 'test@example.org',
+          it('returns the submitter', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: { ...variables, resourceId: 'abusive-user-id' },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  submitter: {
+                    email: 'test@example.org',
+                  },
+                },
               },
-            },
+              errors: undefined,
+            })
           })
-        })
 
-        it('returns a date', async () => {
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              createdAt: expect.any(String),
-            },
+          it('returns a date', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: { ...variables, resourceId: 'abusive-user-id' },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  createdAt: expect.any(String),
+                },
+              },
+              errors: undefined,
+            })
           })
-        })
 
-        it('returns the reason category', async () => {
-          variables = {
-            ...variables,
-            reasonCategory: 'criminal_behavior_violation_german_law',
-          }
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              reasonCategory: 'criminal_behavior_violation_german_law',
-            },
+          it('returns the reason category', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: {
+                  ...variables,
+                  resourceId: 'abusive-user-id',
+                  reasonCategory: 'criminal_behavior_violation_german_law',
+                },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  reasonCategory: 'criminal_behavior_violation_german_law',
+                },
+              },
+              errors: undefined,
+            })
           })
-        })
 
-        it('gives an error if the reason category is not in enum "ReasonCategory"', async () => {
-          variables = {
-            ...variables,
-            reasonCategory: 'my_category',
-          }
-          await expect(action()).rejects.toThrow(
-            'got invalid value "my_category"; Expected type ReasonCategory',
-          )
-        })
+          it('gives an error if the reason category is not in enum "ReasonCategory"', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: {
+                  ...variables,
+                  resourceId: 'abusive-user-id',
+                  reasonCategory: 'category_missing_from_enum_reason_category',
+                },
+              }),
+            ).resolves.toMatchObject({
+              data: undefined,
+              errors: [
+                {
+                  message:
+                    'Variable "$reasonCategory" got invalid value "category_missing_from_enum_reason_category"; Expected type ReasonCategory.',
+                },
+              ],
+            })
+          })
 
-        it('returns the reason description', async () => {
-          variables = {
-            ...variables,
-            reasonDescription: 'My reason!',
-          }
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              reasonDescription: 'My reason!',
-            },
+          it('returns the reason description', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: {
+                  ...variables,
+                  resourceId: 'abusive-user-id',
+                  reasonDescription: 'My reason!',
+                },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  reasonDescription: 'My reason!',
+                },
+              },
+              errors: undefined,
+            })
           })
-        })
 
-        it('sanitize the reason description', async () => {
-          variables = {
-            ...variables,
-            reasonDescription: 'My reason <sanitize></sanitize>!',
-          }
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              reasonDescription: 'My reason !',
-            },
+          it('sanitize the reason description', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: {
+                  ...variables,
+                  resourceId: 'abusive-user-id',
+                  reasonDescription: 'My reason <sanitize></sanitize>!',
+                },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  reasonDescription: 'My reason !',
+                },
+              },
+              errors: undefined,
+            })
           })
         })
-      })
 
-      describe('reported resource is a post', () => {
-        beforeEach(async () => {
-          await factory.create('Post', {
-            author: user,
-            id: 'p23',
-            title: 'Matt and Robert having a pair-programming',
-            categoryIds,
+        describe('reported resource is a post', () => {
+          beforeEach(async () => {
+            await factory.create('Post', {
+              author: currentUser,
+              id: 'post-to-report-id',
+              title: 'This is a post that is going to be reported',
+              categoryIds,
+            })
           })
-          variables = {
-            ...variables,
-            resourceId: 'p23',
-          }
-        })
 
-        it('returns type "Post"', async () => {
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              type: 'Post',
-            },
+          it('returns type "Post"', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: {
+                  ...variables,
+                  resourceId: 'post-to-report-id',
+                },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  type: 'Post',
+                },
+              },
+              errors: undefined,
+            })
           })
-        })
 
-        it('returns resource in post attribute', async () => {
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              post: {
-                title: 'Matt and Robert having a pair-programming',
+          it('returns resource in post attribute', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: {
+                  ...variables,
+                  resourceId: 'post-to-report-id',
+                },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  post: {
+                    title: 'This is a post that is going to be reported',
+                  },
+                },
               },
-            },
+              errors: undefined,
+            })
           })
-        })
 
-        it('returns null in user attribute', async () => {
-          await expect(action()).resolves.toMatchObject({
-            report: {
-              user: null,
-            },
+          it('returns null in user attribute', async () => {
+            await expect(
+              mutate({
+                mutation: reportMutation,
+                variables: {
+                  ...variables,
+                  resourceId: 'post-to-report-id',
+                },
+              }),
+            ).resolves.toMatchObject({
+              data: {
+                report: {
+                  user: null,
+                },
+              },
+              errors: undefined,
+            })
           })
         })
-      })
 
-      /* An der Stelle würde ich den p23 noch mal prüfen, diesmal muss aber eine error meldung kommen.
+        /* An der Stelle würde ich den p23 noch mal prüfen, diesmal muss aber eine error meldung kommen.
            At this point I would check the p23 again, but this time there must be an error message. */
 
```


can we delete this comment? is it still relevant?